### ISSUE TITLE
Missing dependency on typing_extensions in Python 3.9

### ISSUE THREAD

**Issue by paulmelnikow** [python-attrs/attrs#1087]
Importing this package in Python 3.9 without typing_extensions installed causes an ImportError.

Would it make sense to declare a dependency on typing_extensions for Python < 3.10?

**Comment by hynek** [comment]
That's not true:

```console
❯ vf tmp --python python3.9
Creating tempenv-5d862127335c4 via ~/.asdf/shims/python3.9 …

$ pip install attrs
Looking in indexes: https://pypi.vm.ag/root/vrmd/+simple/
Collecting attrs
  Using cached https://pypi.vm.ag/root/pypi/%2Bf/29e/95c7f6778868d/attrs-22.2.0-py3-none-any.whl (60 kB)
Installing collected packages: attrs
Successfully installed attrs-22.2.0

$ python -c "import attrs; print(attrs.__version__)"
22.2.0

$ pip list
Package    Version
---------- -------
attrs      22.2.0
pip        22.3.1
setuptools 65.6.3
wheel      0.38.4
```

Seems like you're accessing the type stubs or something?

We do rely on the fact that once you have Mypy installed, you'll get *typing_extensions* automatically?

**Comment by webknjaz** [comment]
@hynek doesn't it only install typings with an explicit flag?

**Comment by hynek** [comment]
No, that's something different. Mypy currently [depends on typing-extensions](https://github.com/python/mypy/blob/77f872568998bf54a24f497fef72893a60d84633/setup.py#L215).

**Comment by hynek** [comment]
(it's probably not super kosher to depend on that and we might run into problems if we say move away from stubs, but I would like to see a repo for an import error in case there's already a need to act _now_. attrs currently has no dependencies and would kind like to leave it that way)

**Comment by paulmelnikow** [comment]
Hmmm, we're not consuming attrs directly. I think the problem is via importing jsonschema. I wonder if jsonschema accesses the attrs types at runtime.

We use mypy at development time, but it is not installed at runtime, so typing_extensions wasn't being installed.

I don't yet have an easy way to troubleshoot this. What is `vf`, by the way?

**Comment by hynek** [comment]
jsonschema works here too:

```console
$ python -c "import jsonschema; print(jsonschema.__version__)"
<string>:1: DeprecationWarning: Accessing jsonschema.__version__ is deprecated and will be removed in a future release. Use importlib.metadata directly to query for jsonschema's version.
4.17.3

$ python -c "import jsonschema; print(jsonschema.validate({}, {}))"
None

$ pip list
Package    Version
---------- -------
attrs      22.2.0
jsonschema 4.17.3
pip        22.3.1
pyrsistent 0.19.3
setuptools 65.6.3
wheel      0.38.4
```

Any chance you could show us the full traceback?

---

`vf` is virtualfish, a really nice virtualenv manager for the Fish shell: https://virtualfish.readthedocs.io/

**Comment by paulmelnikow** [comment]
That's the thing, I can't get the full traceback. The problem is manifesting in Lambda when the handler function is imported, and all I get is that there's an import error. I narrowed the issue down to attrs by inspecting the source code. It's the only place I could find where typing_extensions would be imported in Python 3.9.

I'll try to get a roundtrip set up to test the zipfile, and will try removing the type stubs. If I still get the import error without the type stubs that would mean the problem is definitively elsewhere.

**Comment by hynek** [comment]
Wow OK, yeah that sounds like a testing nightmare and I guess it's possible lambda could do some weird importing no one asked it of?

Pls keep us posted!

**Comment by hynek** [comment]
I'm gonna close this since there's nothing going on, but would still love to hear if you find anything out.
